### ISSUE TITLE
feature request: Fortinet, pfsense, opnsense integration

### ISSUE THREAD

**Issue by zombiehoffa** [chiefonboarding/ChiefOnboarding#141]
integrate with fortinet/pfsense/opnsense networking gear to create vpn users among other potential user provisioning things.

https://www.fortinet.com/
https://www.pfsense.org/
https://opnsense.org/

**Comment by ropelletier** [comment]
Maybe this is not my place, but I think this is beyond the scope of this project. A task should be created in Chief Onboarding that notifies the IT Department to create these accounts. If the department wants to automate these then that would be up to that department. But I think it would be a mistake to try to automate the creation of the these accounts. The developer can not possibly code this, and maintain the code for all the possible services out there. 

I did suggest creating a webhook task for exactly this type of thing. Then the automation/coding for this is left up to us, which I believe is where these particular services belong. 

That being said, I do agree that O365 Provisioning is a good suggestion.

**Comment by zombiehoffa** [comment]
If  this project is going to go that way t hen I would suggest, at a bare minimum, having a community repository of automations, kind of like the aur for arch linux. a place where users can easily view the code for the automation to verify it does what it says it does while also making it easy to click a few buttons, fill out a few fields and integrate the automation. 

I think it's a huge mistake to not target the whole pipeline of user creation because that is where the labor saving actually is with a product like this. Spending the time to set it up and host it (or pay to host it) solves very little of the actual problem here if the end point is sending an email to the IT department (I might as well just create an email template for each of the 5-10 employee types I have and fill in the information in that email myself and avoid the work of dealing with this app). It's also wasteful to leave automating specific tasks solely to each user of the program because there is huge overlap between most companies (the integrations I suggested are for products used bya huge percentage of businesses for example).

At the same time, I do agree with you that the developers likely don't have the resources to do it all for the thousands of endpoints out there and because of that I think an AUR type solution (or docker hub if you prefer that analogy) would solve both situations because eventually someone who has to integrate something will contribute it back and before long you will have a large library of community integrations (and a dataset of how many people are using which integration, which would be an invaluable guide about what integrations should be brought under the base application for increased support and peace of mind for potential users).

**Comment by GDay** [comment]
@zombiehoffa see also my comment in the other issue, here: https://github.com/chiefonboarding/ChiefOnboarding/issues/140#issuecomment-1088659790

Your comment is a good point. So, what I could do is the following:

1. Implement everything as mentioned in the comment I made (see link above). 
2. Then create an export/import tool to be able to share simple JSON config (basically the endpoints and perhaps some specific configuration towards that - no tokens obviously). 

The JSON config can be shared with others so that setting up integrations will just be a few button clicks. Those integrations will obviously then community based (though I can create some as well). I can set up a forum or so where those can be shared. 

I could set up a repository where users can submit theirs and then quickly pull in their own instance, but I am a little on the fence on connecting anything from me to other instances. I like the way it is now (no instances are connected to me in any way - except the ones I set up myself obviously).

Thoughts?

**Comment by zombiehoffa** [comment]
manually copying over json is fine too, the most important bit as far as user experience is concerned is that whatever is the post with the json is clearly delineated from comments and other stuff (aka you don't want to be the xda developers forum). Second  most important would be searchable. I liked the aur example because the human interface is about as simple as you can get (each "post" is a package) and there are really useful columns like votes and popularity (helps a lot when there is more than one package option), as well as a rudimentry comment section below that is often filled with good ideas to fix issues. I didn't necessarily mean it need to automagically be able to be pulled into chief (even though that would be sweet). I do think that having some way to even just get a tiny idea of what endpoints are popular would be useful to the project because it would highlight what probably should be brought in house with official support/development.


For reference: https://aur.archlinux.org/packages

**Comment by ropelletier** [comment]
@zombiehoffa This was indeed the point I was really trying to make "I do agree with you that the developers likely don't have the resources to do it all for the thousands of endpoints". Which is why I was simply advocating for a method by which we could build our own as needed, again my vote was for webhooks. I do really love your idea of some kind of plugin repository for exactly this purpose where people could build it and then contribute to the community. You mentioned AUR; I think something like the way zabbix and grafana plugins work that just use json style plugins like @GDay was talking about.

I have never developed in django; I really only due PHP, jQuery, javascript, mostly with wordpress. My python is mediocre at best, so I don't think I will be much help regarding contributing. I mention this only to make the point that any improvements you make to the software is appreciated.

**Comment by zombiehoffa** [comment]
@ropelletier The aur is just what came to mind since I run an arch derivative on my desktop it was on my mind. I'll be happy with whatever is done as long as there is a reasonable way for me to contribute back what I create (and benefit from what everyone else has to create to get the endpoints they carea bout integrated). I'll be even happier if whatever we end up with has some rudimentary rating/popularity/commenting system like the aur has to help decide which integration to use and troubleshoot potential problems.

**Comment by GDay** [comment]
@zombiehoffa @ropelletier @gb-kpatel

Would love your feedback on what I have done so far. 

Example JSON (specifically for the Asana api in this case, which is working!):
```
{
	"headers": {
		"Content-Type": "application/json",
		"Accept": "application/json",
		"Authorization": "Bearer $TOKEN"
	},
	"form": [{
		"id": "TEAM_ID",
		"name": "Select team to add user to",
		"url": "https://app.asana.com/api/1.0/organizations/{{ORG}}/teams",
		"type": "choice",
	        "items": "data",
		"choice_id": "gid",
		"choice_name": "name",
		"multiple": true
	}],
	"exists": {
		"method": "GET",
		"url": "https://app.asana.com/api/1.0/users/{{new_hire.email}}/teams?organization={{ORG}}",
		"data": {
			"data": {
				"user": "{{new_hire.email}}"
			}
		},
		"expected": "{{new_hire.email}}"
	},
	"execute": [{
			"method": "POST",
			"url": "https://app.asana.com/api/1.0/workspaces/{{ORG}}/addUser",
			"data": {
				"data": {
					"user": "{{new_hire.email}}"
				}
			}
		},
		{
			"method": "POST",
			"url": "https://app.asana.com/api/1.0/teams/{{TEAM_ID}}/addUser",
			"data": {
				"data": {
					"user": "{{new_hire.email}}"
				}
			}
		}
	],
	"initial_data_form": [{
			"id": "TOKEN",
			"name": "Please put your token here",
			"description": "You can find your token here: https://...."
		},
		{
			"id": "ORG",
			"name": "Organization id",
			"description": "You can find your organization id here: https://..."
		}
	]
}
```

Let's go through each of the items:

`initial_data_form`: Once this JSON is imported in the integrations section, the user will get a form based on the items in this array. In this case, we need the api token and the organization id to make API calls to Asana. The generated form looks like this:
![image](https://user-images.githubusercontent.com/1939656/162969584-10a27b80-e6aa-4929-956a-a3039abfc267.png)


`headers`: The headers that are sent with each request. Generally authentication and such.

`form`: This is the form that will be displayed in the sequences area. I have moved all integrations to the sequences, so you can drop an integration into a sequence (to trigger when a user has completed something or based on a specific workday/time). You could create a sequence in which the user will have access to a platform after they completed their first week for example. Or perhaps elevate permissions after a specific amount of time.
The form will be shown when you drop this integration into the sequence. You can basically quickly customize the behavior of the integration/webhook here. For my asana account it will look like this:

![image](https://user-images.githubusercontent.com/1939656/162970324-09397b93-26fa-463f-b2a9-a15c2b76dcb8.png)
It shows the teams that are in my account (again see the data that I put in the `form` area of the JSON), currently fetched from a url, but it could also be hardcoded options. 

`exists`: if you are using a simple webhook, then you do not need this one, but for account provisioning, you will likely want to have this. With this item, you can check if a user already exists in the team. This will be shown on the `access` page specific to the new hire:

![image](https://user-images.githubusercontent.com/1939656/162970758-cdd18f6c-8e9f-4d08-b10b-c0ddef25efc6.png)
From here, you can trigger the `execute` area if you would want to create the account manually.

`execute`: whatever it needs to do to complete the process. This could be one API call, but also multiple. For Asana, two calls are necessary. 

In all areas, you can use specific data from the new hire, such as email, first name, last name, position... And you can also use the variables that you require with `initial_data_form` and the ones the admin provided through the `form`.

There are a few bugs that I am still working on, but overall, it looks pretty solid. Though, it's not set in stone yet, happy to hear your thoughts!

**Comment by GDay** [comment]
Webhooks are released - OAuth2 is still in the works (but should work sometime next week). Docs: https://docs.chiefonboarding.com/integrations/

@zombiehoffa regarding the user repository, I decided to go with Flarum, this is open source forum software. Every topic is an integration. People can vote and leave comments on each integration. It's live: https://integrations.chiefonboarding.com/

Regarding the integrations mentioned here:
**Fortinet** - uses basic auth which should be possible with what has been implemented so far.
**pfsense** - https://github.com/jaredhendrickson13/pfsense-api - needs this plugin I think for an API. That API allows a token, which should work with what we have right now.
**opnsense** - https://docs.opnsense.org/development/api.html - uses basic auth as well, so should be possible.

I am open to suggestions/feedback!
